osf/models/mixins.py

```diff
@@ -335,6 +347,41 @@ def remove_affiliated_institution(self, inst, user, save=False, log=True):
             return True
         return False
 
+    def update_institutional_affiliation(self, auth, institution_ids):
+        user = auth.user
+        current_institutions = set(self.affiliated_institutions.values_list('_id', flat=True))
+
+        institutions_to_add = set(institution_ids) - current_institutions
+        institutions_to_remove = current_institutions - set(institution_ids)
+
+        Institution = apps.get_model('osf.Institution')
+
+        from osf.exceptions import UserNotAffiliatedError
+
+        for institution_id in institutions_to_add:
+            try:
+                institution = Institution.objects.get(_id=institution_id)
+                self.add_affiliated_institution(institution, user, log=True)
+            except Institution.DoesNotExist:
+                raise ValidationError(f'User is not affiliated with {institution.name},'
+                                             f' it was not found in records')
+            except UserNotAffiliatedError:
+                raise ValidationError(f'User is not affiliated with {institution.name},')
+
+        for institution_id in institutions_to_remove:
+            try:
+                institution = Institution.objects.get(_id=institution_id)
+                self.remove_affiliated_institution(institution, user, log=True)
+            except Institution.DoesNotExist:
+                raise ValidationError(f'User is not affiliated with {institution.name},'
+                                             f' it was not found in records')
+            except UserNotAffiliatedError:
+                raise ValidationError(f'User is not affiliated with {institution.name},')
+
+        self.save()
+
+        self.update_search()
```

Comment:
This seems more complicated than necessary since both `add_affiliated_institution()`and ` remove_affiliated_institution()` handles already affiliated or not affiliated cases. No need to make a all query and then calculate the diff set.

In addition, is their any special consideration for not passing a list of institution objects? This will save us all the extra queries to find the institution by ID.